### PR TITLE
Mdd merge spamgourmet clone

### PR DESCRIPTION
**Pull request by mikedlr**
The aim of this pull request is to merge the spamgourmet-clone repository into the spamgourmet repository which should allow us to improve the deployment and then testing of spamgourmet, making it practical to maintain and also making it easy to run clone services from.  

I've got to the stage of being able to create a Docker container that contains the spamgourmet install, including having the database scripts work, however I haven't yet ran the system from that container.

### CODE REVIEW

File: spamgourmet-clone/scripts/install-spamgourmet.sh [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+	fi
+	/var/lib/dehydrated/renewAllCerts.sh
+	service exim4 restart
+	service lighttpd restart
+}
+
+function configure_sg_db_connection_and_data {
+	#ensure that mysqld is running, - what about init systems?
+	#    mysqld_safe --no-watch
+	service mysql start
```
Comment (eegeeZA) [inline spamgourmet-clone/scripts/install-spamgourmet.sh:258]: Will the DB files be stored outside the Docker image?

File: spamgourmet-clone/scripts/install-sg-exim.sh [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+echo '========================= spamgourmet server installation start'
+source sg-server-config.sh
+
+cd $SCRIPT_BASE_DIR
+
+function exim4_packages {
+	echo '##########################################################################'
+	echo '### install mandatory packages'
+	echo '##########################################################################'
+	apt-get update; apt-get upgrade; apt-get autoremove
```
Comment (eegeeZA) [inline spamgourmet-clone/scripts/install-sg-exim.sh:12]: Most sources have these `RUN` in the Dockerfile directly creating the new layers. Might be worth investigating later the build performance or having a multi-stage build to get a smaller final image size.

**Review by eegeeZA — APPROVED** [review]
Looks good, but I wouldn't mind a few more eyes on this PR. Going to include a workflow to build the image (without pushing anywhere) to check for any issues.

File: spamgourmet-clone/scripts/install-spamgourmet.sh (hunk 1 above)
Comment (mikedlr) [reply spamgourmet-clone/scripts/install-spamgourmet.sh:258]: Eventually we must make that possible. Right now I don't care as long as I can get at least one test running.

File: spamgourmet-clone/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,8 @@
+FROM perl:5.30.3-slim-threaded-buster
+ADD remote-install.sh scripts sg-server-config.sh /
+ADD scripts/* scripts/
+RUN apt-get update && apt-get install -y --no-install-recommends apt-utils
+ENV SCRIPT_BASE_DIR=/ 
+RUN bash -vx /scripts/install-sg-mariadb.sh
+RUN bash -vx /scripts/install-sg-exim.sh
+RUN bash -vx /scripts/install-spamgourmet.sh
```
[the remark this answers is not in the document]
Comment (mikedlr) [reply spamgourmet-clone/Dockerfile:8]: Not sure yet.  I think to be "appropriate" it should really be split into multiple ones with the default entry point in the main one running spamgourmet with various options possible e.g. to set up the domain correctly.

File: spamgourmet-clone/sg-server-config.sh [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+EATENLOGCOUNT=10
+DKIM_SELECTOR="20180125"
+
+OVHLETSENCRYPT=1
+
+if [ $OVHLETSENCRYPT -eq 1 ]; then
+	OVHAPPKEY=CVvttJNslWX7jp0q
+	OVHAPPSECRET=oJHAufRp8EHamKo64d7ClgWpXRJoodUB
+	OVHCONSKEY=ulBZt16KjNp6PmOvgsge8dudNoEm2ZSO
+fi
```
[the remark this answers is not in the document]
Comment (mikedlr) [reply spamgourmet-clone/sg-server-config.sh:22]: I think in a separate PR I will implement precommit to ensure this and then fix any files that are broken at the time. In the meantime I'm messing with [vscodium ](https://vscodium.com/) so please expect some junk. I will clean up a bit.

File: spamgourmet-clone/scripts/config-lets-encrypt.sh [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+### domains.txt file
+##########################################################################
+
+echo '##########################################################################'
+echo "### configure dehydrated for $DOMAIN with DNS managed by OVH"
+echo '##########################################################################'
+apt-get install -y git python-pip
+mkdir -p /var/lib/dehydrated/hooks
+cd /var/lib/dehydrated/
+git clone https://github.com/rbeuque74/letsencrypt-ovh-hook hooks/ovh
```
[the remark this answers is not in the document]
Comment (mikedlr) [reply spamgourmet-clone/scripts/config-lets-encrypt.sh:25]: a fork/clone is also a security risk in that it won't get security updates - A good solution would be to download on a checksum (git commit hash is acceptable though really shouldn't be SHA1) and crash out for manual verification before doing a final release if the code changes.

File: spamgourmet-clone/scripts/install-sg-mariadb.sh [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+source sg-server-config.sh
+
+cd $SCRIPT_BASE_DIR
+
+function mysql_packages {
+	echo '##########################################################################'
+	echo '### install mandatory packages'
+	echo '##########################################################################'
+	apt-get update; apt-get upgrade; apt-get autoremove
+	apt-get install -y mariadb-server \
```
[the remark this answers is not in the document]
Comment (mikedlr) [reply spamgourmet-clone/scripts/install-sg-mariadb.sh:13]: Yes, I'd love that. I'm not sure if we should swap or have both. Current docker config has the advantage of matching the spamgourmet actual configuration, however Josiah sounds keen on changing that so if we brought that along we would be able to stop doing as much custom work on mariadb, which is a definite advantage.

File: spamgourmet-clone/Dockerfile (hunk 3 above)
[the remark this answers is not in the document]
Comment (mikedlr) [reply spamgourmet-clone/Dockerfile:8]: Each of the run scripts is quite long, so at least whilst we're developing this and keep having to rerun the scripts, the extra layers cache some of the work and reduce the build time.  Let's change this when we see this working stably, but not now.

File: spamgourmet-clone/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,8 @@
+FROM perl:5.30.3-slim-threaded-buster
+ADD remote-install.sh scripts sg-server-config.sh /
+ADD scripts/* scripts/
```
[the remark this answers is not in the document]
Comment (mikedlr) [reply spamgourmet-clone/Dockerfile:3]: Thanks. Will fix.

File: spamgourmet-clone/scripts/config-lets-encrypt.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (eegeeZA) [reply spamgourmet-clone/scripts/config-lets-encrypt.sh]: Yes, that sounds good. So something like [checking out the commit hash](https://stackoverflow.com/a/3555202) and then [comparing with upstream](https://stackoverflow.com/a/17938274) to see if we're behind?

**Comment by mikedlr** [comment]
I'm happy with this as it is now. It doesn't touch the main spamgourmet code so there is no risk to that and, with @vasile-gh's testing it is now much closer to a good way to deliver an installed spamgourmet clone.  Also the docker container continues to build so we have some kind of minimal check that the code is sane. 

I'm creating a proper merge commit since this represents the bringing together of the histories of two separate repositories and I want to keep the correct development history up to this stage.

**Pull request merged**